tsdate/variational.py

```diff
@@ -563,32 +595,106 @@ def iterate(
             self.scale,
         )
 
-        # for debugging
-        # assert self._check_valid_state(
-        #    self.parents, self.children, self.posterior,
-        #    self.node_factors, self.edge_factors,
-        # )
+        if check_valid:  # for debugging
+            assert self._check_valid_state(
+                self.parents,
+                self.children,
+                self.posterior,
+                self.node_factors,
+                self.edge_factors,
+            )
 
         return np.nan  # TODO: placeholder for marginal likelihood
 
+    def rescale(
+        self,
+        *,
+        rescale_intervals=1000,
+        rescale_segsites=False,
+        use_median=False,
+        quantile_width=0.5,
+    ):
+        """Normalise posteriors so that empirical mutation rate is constant"""
+        edge_weights = (
+            np.ones(self.edge_weights.size) if rescale_segsites else self.edge_weights
+        )
+        nodes_time = self._point_estimate(self.posterior, self.constraints, use_median)
+        original_breaks, rescaled_breaks = mutational_timescale(
+            nodes_time,
+            self.likelihoods,
+            self.constraints,
+            self.parents,
+            self.children,
+            edge_weights,
+            rescale_intervals,
+        )
+        self.posterior[:] = piecewise_scale_posterior(
+            self.posterior,
+            original_breaks,
+            rescaled_breaks,
+            quantile_width,
+            use_median,
+        )
+        self.mutations_posterior[:] = piecewise_scale_posterior(
+            self.mutations_posterior,
+            original_breaks,
+            rescaled_breaks,
+            quantile_width,
+            use_median,
+        )
+
     def run(
         self,
         *,
-        ep_maxitt=20,
-        em_maxitt=10,
+        ep_maxitt=10,
         max_shape=1000,
         min_step=0.1,
-        min_kl=True,
-        progress=None
+        min_kl=False,
+        rescale_intervals=1000,
+        rescale_segsites=False,
+        regularise=True,
+        progress=None,
     ):
-        for itt in tqdm(
+        nodes_timing = time.time()
+        for _ in tqdm(
             np.arange(ep_maxitt),
             desc="Expectation Propagation",
             disable=not progress,
         ):
             self.iterate(
-                em_maxitt=em_maxitt if itt else 0,
                 max_shape=max_shape,
                 min_step=min_step,
                 min_kl=min_kl,
+                regularise=regularise,
+            )
+        nodes_timing -= time.time()
+        skipped_nodes = np.sum(np.isnan(self.log_partition))
+        if skipped_nodes:
+            logging.info(f"Skipped {skipped_nodes} nodes with invalid posteriors")
+        logging.info(f"Calculated node posteriors in {abs(nodes_timing)} seconds")
```

Comment:
Similarly here - maybe a tqdm progress bar would be better than reporting timings?

Reply:
I prefer logging, not a fan of progress bars, and they won't work with numba anyway